### PR TITLE
Test

### PR DESCRIPTION
**Pull request by Hoijof**

### CODE REVIEW

File: src/game/GameWorld.ts [hunk 1]
```diff
@@ -49,7 +49,7 @@ export default abstract class GameWorld {
     return;
   }
   onError(e: Error): void {
-    return;
+    throw e;
```
Comment (Hoijof) [inline src/game/GameWorld.ts:52]: Details

File: src/game/entities/Player/Player.ts [hunk 2]
```diff
@@ -31,8 +31,8 @@ export default class Player extends PlayerSerializable {
   }
 
   basicUpdate(basicUpdate: PlayerBasicUpdate): void {
-    this.x = basicUpdate[1];
-    this.y = basicUpdate[2];
+    super.basicUpdate([basicUpdate[0], basicUpdate[1], basicUpdate[2]]);
```
Comment (Hoijof) [inline src/game/entities/Player/Player.ts:34]: Matias halluda

File: src/game/entities/Projectile/ProjectileDrawer.ts [hunk 3]
```diff
@@ -7,12 +7,12 @@ export default class ProjectileDrawer extends Drawable {
...
   draw(camera: Camera): void {
-    super.draw(camera);
-
     this.context.fillStyle = this.e.isColliding ? PALETTE.black : PALETTE.red;
 
     this.context.beginPath();
     this.context.arc(this.e.x, this.e.y, this.e.r, 0, 2 * Math.PI, false);
     this.context.fill();
+
+    super.draw(camera);
```
Comment (Hoijof) [inline src/game/entities/Projectile/ProjectileDrawer.ts:16]: We want to draw the debug data on top of entities, not behind

File: src/game/entities/Serializable.ts [hunk 4]
```diff
@@ -4,10 +4,18 @@ import {
...
+import {
+  applyBasicUpdateTolerance,
+  getEntityFromType,
+  toFixedIfNecessary,
+} from '../utils';
 import GameObject from './GameObject';
 
 export default abstract class Serializable extends GameObject {
+  //#region Debug
+  maxXTolerance = 0;
```
Comment (Hoijof) [inline src/game/entities/Serializable.ts:16]: This field is the max difference between the client x and the received server X (on a basic update) on any given frame.

File: src/game/drawing/debug.ts [hunk 5]
```diff
@@ -31,28 +32,38 @@ export function drawPlayerAcceleration(player: Player): void {
...
-  context.fillText(`Acc: ${player.ax} / ${player.ay}`, x, y + fontPadding * 3);
   context.fillText(
-    `Speed: ${player.vx} / ${player.vy}`,
+    `Speed: ${entity.vx} / ${entity.vy}`,
     x,
-    y + fontPadding * 4
+    y + fontPadding * 3
   );
+
+  if (entity instanceof Player) {
```
Comment (amatiasq) [inline src/game/drawing/debug.ts:60]: note to self: use type guards

File: src/game/entities/Player/Player.ts (hunk 2 above)
Comment (amatiasq) [reply src/game/entities/Player/Player.ts:34]:
```seggestion
super.basicUpdate(basicUpdate.slice(0, 3))
```
o.o?

File: src/game/utils.ts [hunk 6]
```diff
@@ -44,6 +44,15 @@ export function getKeyToIndexFromType(
   return getEntityFromType(type).keyToIndex;
 }
 
+export function applyBasicUpdateTolerance(
+  prevValue: number,
+  newValue: number
+): number {
```
Comment (amatiasq) [inline src/game/utils.ts:50]: get used to implicit return types

**Review by amatiasq — APPROVED** [review]
LCDTH

File: src/game/entities/Serializable.ts [hunk 7]
```diff
@@ -41,8 +51,13 @@ export default abstract class Serializable extends GameObject {
 
   basicUpdate(basicUpdate: PlayerBasicUpdate | CircleBasicUpdate): void;
   basicUpdate(basicUpdate: unknown): void {
-    this.x = basicUpdate[1];
-    this.y = basicUpdate[2];
+    this.maxXTolerance =
+      this.x - basicUpdate[1] > this.maxXTolerance
+        ? this.x - basicUpdate[1]
+        : this.maxXTolerance;
```
Comment (amatiasq) [inline src/game/entities/Serializable.ts:57]:
```suggestion
    this.maxTolerance = Math.max(this.maxXTolerance , this.x - basicUpdate[1])
```

Comment (amatiasq) [inline src/game/entities/Serializable.ts:57]:
```suggestion
    this.maxTolerance = Math.max(this.maxXTolerance, this.x - basicUpdate[1])
```

**Review by amatiasq — CHANGES_REQUESTED** [review]
PNTVRDRALPMQTRMPYLPPSPPUSDMCQDEQTTG

File: src/game/entities/Serializable.ts (hunk 4 above)
Comment (amatiasq) [reply src/game/entities/Serializable.ts:16]: then the name is misleading
```suggestion
  maxDeviationRegistered = 0;
```

**Pull request merged**